staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go

```diff
@@ -53,6 +54,8 @@ type RequestScope struct {
 	Typer           runtime.ObjectTyper
 	UnsafeConvertor runtime.ObjectConvertor
 
+	OpenAPISpecGetter openapi.SpecGetter
```

Comment:
This type holds enough information to serve `/apis/group/version/namespaces/name/resource/***`.  The openapi is not served from this location.  This doesn't look like the right place to add an openapi getter.

Also, this type is focused on being able to serve a single apiresource and as I recall the openapi is based around specifying it for an entire suite of apigroups (not even just a particular version).

Can you elaborate on why this is needed for a single api resource and then break up the openapi generation to be scoped to what you need instead of broadly including all of it?

Reply:
> Also, this type is focused on being able to serve a single apiresource and as I recall the openapi is based around specifying it for an entire suite of apigroups (not even just a particular version)

The serializer/convert[o|e]r/etc all know about multiple objects unless we've fixed that?

I asked for only the data model to be passed around, but @mbohlool says that even though there's a static data model linked in, it will still need to be re-parsed (e.g. when we hook up the extensions-apiserver, at least). I think I'm still fine with just passing the staticly linked data models and fixing the code for CRDs when we allow CRDs, but I don't really care that much.

The thing I care about is accidental cross-resource references, mainly from CR -> built in types. I don't want those in until we're ready to be deliberate about it.

I looked with Jenny and there's not a clearly better place to put this than the request scope, since it needs to end up in multiple handlers (PUT, POST, PATCH). I think it might be reasonable to add a layer to filter out all operations except for the one in question, but it's not obvious to me that the benefit is worth the extra effort (which shouldn't be too much since we already do this filtering after adding the operations in the first place, but still: what's the benefit?).

> Can you elaborate on why this is needed for a single api resource

It's needed for
* Merging (only apply)
* Validation automation (any mutating request)

Reply:
> Merging (only apply)

Looks specific to the feature branch.  I'm not seeing why we'd merge this piece into master.  This isn't a general improvement, it's a new piece of plumbing that you only need in the feature branch.  

> Validation automation (any mutating request)

I don't recognize this feature.  I don't think a feature we don't have is a reason for the change in master.

If you want to have a separate feature for it in master, that's certainly a piece we can break off and talk about adding.  It's much smaller than apply and I think most of the concern would be around compatibility and tightly bounding the information about openapi.

Reply:
I think there is value in keeping the delta between master & feature branch low. I think the risk from a future rebase problem is higher than the risk of including in master a thing that's not hooked up yet. (It does need an explanatory test, though, or it can't be expected to stick around)

Reply:
@deads2k 
Would you feel more comfortable if what was exposed to the handlers through the RequestScope only allowed them to look up openapi schema for types within the same api group? Or is your issue more with the internal/external type boundary.